### PR TITLE
Analyze model results

### PR DESCRIPTION
**Pull request by mangomadhava**
## Changes: 

**Setting up a labeling task.ipynb**: Changed to use the database csv instead of the firestore when querying for new images. 

## New files: 

**analyze_model_output.ipynb**: Displays the model results by each region and crop. The first few cells in the notebook explain how to use it. It runs the [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) function from YOLO library on the images in the current dataset separated by region, and then displays class performance within each region. 

**query_uploaded_database.ipynb**: User selects if they want to find images from a region/crop/crop+region. It will pull random images that are not in test set into the local file system and run [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) on it. The user is able to modify the conf-threshold and iou threshold if enough images are not being found. The images are then uploaded as a dataset on google cloud. 

**split_labels.ipynb**: Takes the currently labeled images on google cloud datasets and creates a random 80-20 split stratified by country. The images are then uploaded as a downloadable dataset on google cloud in the street2sat-gcloud-labeling folder to be used for future training. 

**test_set_creation.ipynb**: Randomly samples images by country from the database-info.csv (located in street2sat-database-csv bucket) which are not being labeled and creates a test-set.csv which is added to street2sat-database-csv bucket. 

**get_original_images.py**: helper script to copy images from one bucket in google cloud to another for street2sat. 

**run_on_folder.py**: helper script to run yolo detections on a folder on the local machine 

**update_csv_from_firebase.py**: This script updates the database-info.csv located inside 'street2sat-database-csv' bucketwith the associated predictions and changes from the firestore.

### CODE REVIEW

File: notebooks/Setting up a labeling task.ipynb [hunk 1]
```diff
@@ -128,18 +128,14 @@
     "\n",
```
Comment (ivanzvonkov) [inline notebooks/Setting up a labeling task.ipynb:109]: <p>Can we now use the csv here too?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (ivanzvonkov) [inline notebooks/Setting up a labeling task.ipynb:109]: <p> Is it possible to remove the entire dependency on firestore here and completely depend on the csv?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (ivanzvonkov) [inline notebooks/Setting up a labeling task.ipynb:109]: <pre class="ql-syntax" spellcheck="false">Line #6.        # need to double check if these changes work!!!!!
</pre><p>Was this checked?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: notebooks/analyze_model_output.ipynb [hunk 2]
```diff
@@ -0,0 +1,645 @@
+{
```
Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <p>A format that I have started to follow in other notebooks is:</p><h1>Title</h1><p><strong>Author</strong>: &lt;Name&gt; (&lt;email&gt;)</p><p><strong>Description:</strong>&nbsp;&lt;what you wrote in the PR&gt;</p><p> </p><p>Do you think it would be useful to add to notebooks like this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <p>Is firestore connection still necessary here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <ol><li>Where does this yaml config file come from? </li><li>The validation script needs to be run on training data too or am I missing something?</li><li>Ok so list of images -&gt; config yaml -&gt; model validation -&gt; txt output, is that correct? If yes, do we need to get the list of images from txt output and not just use the original list of images?</li><li>Can this be done using csv?</li><li>Is running the validation script necessary for each country, don't we already have all the predictions for each country just together?</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #10.    yaml_file = os.path.join(folder, dataset_name, 'data_info.yaml')
</pre><p>Is data_info.yaml the right name or config.yaml?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #13.    model_weights_path = '/gpfs/data1/cmongp1/mpaliyam/street2sat/yolov5/runs/train/exp18/weights/best.pt'  
</pre><p>Not relevant for this PR, but for future it would be good to track the weights in dvc also!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (ivanzvonkov) [inline notebooks/analyze_model_output.ipynb]: <p>Does it have to be a link to that specific commit?</p><p> </p><p>Also, it might make sense to tie this heading to the numbered outline above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

File: notebooks/query_uploaded_database.ipynb [hunk 3]
```diff
@@ -0,0 +1,648 @@
+{
```
Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <p>Same comment about title, author, description in previous notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <p>Ok so if I understand correctly this is another way to create labeling tasks right? Would it make sense to combine this with the other notebook so we don't have diverging dataset creation methods?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <pre class="ql-syntax" spellcheck="false">data = yaml_file # dataset.yaml path
imgsz= 640&nbsp; # inference size (height, width)
conf_thres = 0.05&nbsp; # confidence threshold
iou_thres = 0.05&nbsp; # NMS IOU threshold
max_det = 1000&nbsp; # maximum detections per image
project = 'runs/detect'&nbsp; # save results to project/name
name ='exp'&nbsp; # save results to project/name
</pre><blockquote>Don't be afraid to make a name long. A long descriptive name is better than a short enigmatic name. A long descriptive name is better than a long descriptive comment.</blockquote><blockquote><strong>Robert C. Martin (Clean Code)</strong></blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <p>Generally, I think it's best to clear outputs of all notebooks before storing on Github</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <p>I think this is redundant</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ivanzvonkov) [inline notebooks/query_uploaded_database.ipynb]: <p>I think this can be removed</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: notebooks/split_labels.ipynb [hunk 4]
```diff
@@ -9,7 +9,7 @@
   },
```
Comment (ivanzvonkov) [inline notebooks/split_labels.ipynb]: <p>Can the commented lines just be removed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ivanzvonkov) [inline notebooks/split_labels.ipynb]: <p>Long term would it make sense to also just create a csv to avoid interfacing with these jsonl files too much?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (ivanzvonkov) [inline notebooks/split_labels.ipynb]: <p>Would it make sense to store this in the repo?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

File: notebooks/test_set_creation.ipynb [hunk 5]
```diff
@@ -0,0 +1,7095 @@
+{
```
Comment (ivanzvonkov) [inline notebooks/test_set_creation.ipynb]: <p>Best not make the token publicly viewable, I've used https://pypi.org/project/python-dotenv/ for getting around this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/street2sat/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

File: scripts/get_original_images.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+'USA/2021-08-20-croptour/G0051226.JPG', 'USA/2021-08-20-croptour/G0051229.JPG', 'USA/2021-08-20-croptour/G0051251.JPG', 
+'USA/2021-08-20-croptour/G0051418.JPG', 'USA/2021-08-20-croptour/G0051517.JPG', 'USA/2021-08-20-croptour/G0051534.JPG', 
+'USA/2021-08-20-croptour/G0051724.JPG', 'USA/2021-08-20-croptour/G0051794.JPG', 'USA/2021-08-20-croptour/G0051851.JPG', 
+'USA/2021-08-20-croptour/G0051857.JPG', 'USA/2021-08-20-croptour/G0052077.JPG', 'USA/2021-08-20-croptour/G0052160.JPG', 
+'USA/2021-08-20-croptour/G0052175.JPG', 'USA/2021-08-20-croptour/G0052187.JPG', 'USA/2021-08-20-croptour/G0052194.JPG', 
+'USA/2021-08-20-croptour/G0052208.JPG', 'USA/2021-08-20-croptour/G0052303.JPG', 'USA/2021-08-20-croptour/G0052353.JPG', 
+'USA/2021-08-20-croptour/G0052419.JPG', 'USA/2021-08-20-croptour/G0052471.JPG', 'USA/2021-08-20-croptour/G0052539.JPG', 
+'USA/2021-08-20-croptour/G0052546.JPG', 'USA/2021-08-20-croptour/G0052559.JPG', 'USA/2021-08-20-croptour/G0052585.JPG', 
+'USA/2021-08-20-croptour/G0052658.JPG', 'USA/2021-08-20-croptour/G0052687.JPG', 'USA/2021-08-20-croptour/G0063016.JPG', 
+'USA/2021-08-20-croptour/G0063028.JPG']
```
Comment (ivanzvonkov) [inline scripts/get_original_images.py:90]: what's this text for?

File: scripts/get_original_images.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+from google.cloud import storage
+from tqdm import tqdm
+import json 
+
+'''
+This script will create a new bucket on google cloud. 
+
+The input folder/bucket will be used to pull the same named images
+from the street2sat uploaded bucket and copy into the new bucket. 
+'''
```
Comment (ivanzvonkov) [inline scripts/get_original_images.py:12]: What's the exact use case for this script?

File: scripts/run_on_folder.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        '--save-crop '\
+        '--save-txt '\
+        '--save-conf '\
+        '--conf-thres .15 '\
+        '--iou-thres .15 '\
+        '--project /gpfs/data1/cmongp1/mpaliyam/street2sat/data/NACCRI-Catherine/{0} '\
+        '--name {1}'\
+        
+
+l = [('stereo1', '114GOPRO'),('stereo1', '115GOPRO'),('stereo1', '116GOPRO'),
```
Comment (ivanzvonkov) [inline scripts/run_on_folder.py:20]: A more descriptive name would be useful here

File: scripts/run_on_folder.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import os 
+
+
+'''
+Script to easily run detections on folder. Modify as needed.
```
Comment (ivanzvonkov) [inline scripts/run_on_folder.py:5]: I don't quite understand this, what exactly is the input and output?

File: scripts/update_csv_from_firebase.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+the performance of street2sat over time using the past predictions. The firebase would
+only store the latest predictions. 
+
+'''
+
+
+
+
+def get_images_already_being_labelled():
+    """Gets images already labelled"""
```
Comment (ivanzvonkov) [inline scripts/update_csv_from_firebase.py]: nit: IMO labeled is the right (american) spelling, I know we use labelled elsewhere though haha

File: scripts/update_csv_from_firebase.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        else: 
+            print(f"Duplicate detected: {d_image} {i}")
+
+        time_d.append(d_image['time'])
+        focal_length.append(d_image['focal_length'])
+        pixel_height.append(d_image['pixel_height'])
+
+        # weird extra folder case 
+        if d_image['input_img'].split('/')[3].startswith('2021'):
+            country.append('NA')
```
Comment (ivanzvonkov) [inline scripts/update_csv_from_firebase.py]: why not `None`?

File: scripts/update_csv_from_firebase.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
+from google.cloud import firestore
+from google.cloud import storage
+from tqdm import tqdm
+import pandas as pd
+pd.options.display.max_colwidth = 600
+pd.options.display.max_columns = None
+pd.options.display.max_rows = None
+import collections
+import time
```
Comment (ivanzvonkov) [inline scripts/update_csv_from_firebase.py]: time is not needed

File: scripts/update_csv_from_firebase.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+dup_checker = set()
+i = 0
+
+# query limit amount bc queries can't live for more than 30 seconds 
+limit_amount = 10000
+
+test_num = 0
+nones_detected = 0
+
+docs = coll.order_by('input_img').limit(limit_amount).stream()
```
Comment (ivanzvonkov) [inline scripts/update_csv_from_firebase.py:103]: This is not at all necessary, however, I think it might be cleaner to put all the data into a data frame right away and do operations directly on the data frame:

For example
```
all_data = []
for image in docs:
    all_data.append( image.to_dict() )

df = pd.DataFrame(all_data)

# Remove coords with none, etc 
```
Up to you if you want to take a look at this

**Comment by ivanzvonkov** [comment]
7 days :( sorry for taking so long on this!

File: notebooks/Setting up a labeling task.ipynb (hunk 1 above)
Comment (mangomadhava) [reply notebooks/Setting up a labeling task.ipynb:109]: <p>eliminate all dependency on firestore and buckets </p>

Comment (mangomadhava) [reply notebooks/Setting up a labeling task.ipynb:109]: <p>yup</p>

Comment (mangomadhava) [reply notebooks/Setting up a labeling task.ipynb:109]: <p>eventually </p>

File: notebooks/analyze_model_output.ipynb (hunk 2 above)
Comment (mangomadhava) [reply notebooks/analyze_model_output.ipynb]: <p>good idea</p>

Comment (mangomadhava) [reply notebooks/analyze_model_output.ipynb]: <p>nope not needed</p><p> </p>

File: notebooks/split_labels.ipynb (hunk 4 above)
Comment (mangomadhava) [reply notebooks/split_labels.ipynb]: <p>call this "prepare dataset for object detection'</p>

Comment (mangomadhava) [reply notebooks/split_labels.ipynb]: <p>train object detection csv (tracked by dvc) </p><p>val object detection csv (tracked by dvc)</p><p> </p><ul><li>gs uri </li><li>bboxes </li><li>train/val id </li><li>attributes from datasets </li></ul>

**Comment by ivanzvonkov** [comment]
Thanks for updating the notebooks to be ordered and organized, dvc integration looks good too!

**Review by ivanzvonkov — APPROVED** [review]
